### PR TITLE
Extend ApiRx from abstract Base (prepare for promise)

### PR DESCRIPTION
**Pull request by jacogr**
- Extend ApiRx from Base
- Fix Metadata decoding for new Substrate (additional length prefix)
- Read module prefixes from new outer index

### CODE REVIEW

File: packages/api/src/Base.ts [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+   * ```
+   */
+  get tx (): T {
+    assert(!isUndefined(this._extrinsics), INIT_ERROR);
+
+    return this._extrinsics as T;
+  }
+
+  private init (): void {
+    let isReady: boolean = false;
```
Comment (amaurymartiny) [inline packages/api/src/Base.ts]: Is this redundant to the isReady property? One idea is to have isReady as an Observable on this abstract class, and in the Promise implementation, add `get isReady() { return this._isReady.toPromise() }`

File: packages/api/src/Base.ts [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+  method: string,
+  section: string,
+  toJSON: () => any
+};
+
+const l = logger('api');
+
+const INIT_ERROR = `Api needs to be initialised before using, listen on 'ready'`;
+
+export default abstract class ApiBase<R, S, T> extends E3.EventEmitter implements ApiBaseInterface<R, S, T> {
```
[the remark this answers is not in the document]
Comment (jacogr) [reply packages/api/src/Base.ts]: Erk, will fix.

File: packages/api/src/Base.ts [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+  private _extrinsics?: T;
+  private _genesisHash?: Hash;
+  private _storage?: S;
```
[the remark this answers is not in the document]
Comment (jacogr) [reply packages/api/src/Base.ts]: +1

File: packages/api/src/Base.ts [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+   *   api.rpc.newHead().subscribe((header) => {
+   *     console.log(`new block #${header.blockNumber.toNumber()}`);
+   *   });
+   * });
+   * ```
+   */
+  constructor (wsProvider?: WsProvider) {
+    super();
+
+    this._rpcBase = new RpcRx(wsProvider);
```
[the remark this answers is not in the document]
Comment (jacogr) [reply packages/api/src/Base.ts]: Ok, that is supposed to be new Rpc

File: packages/api/src/Base.ts (hunk 1 above)
Comment (jacogr) [reply packages/api/src/Base.ts]: I went down that route, I think it is easier to just emit the 'ready' and the extention can do what it wants. Will have to see with promise, maybe you are right here, I'm not 100% either way.

**Pull request merged**